src/mongo/s/stats/joom_top.cpp

```diff
@@ -0,0 +1,96 @@
+#define MONGO_LOGV2_DEFAULT_COMPONENT ::mongo::logv2::LogComponent::kDefault
+
+#include "mongo/platform/basic.h"
+
+#include "mongo/db/curop.h"
+#include "mongo/db/jsobj.h"
+#include "mongo/db/service_context.h"
+
+#include "mongo/s/stats/joom_top.h"
+
+namespace mongo {
+
+namespace {
+
+const auto getJoomTop = ServiceContext::declareDecoration<JoomTop>();
+
+}  // namespace
+
+// static
+JoomTop& JoomTop::get(ServiceContext* service) {
+    return getJoomTop(service);
+}
+
+void JoomTop::record(OperationContext* opCtx, bool isError) {
+    auto& curOp = *CurOp::get(opCtx);
+    auto ns = curOp.getNS();
+
+    if (ns[0] == '?' || curOp.getNSS().isCommand()) {
+        return;
+    }
+
+    auto elapsedMicros = opCtx->getElapsedTime().count();
+    auto cmdName = mongoCommandToCommandName(curOp.getCommand());
+
+    auto hashedNs = UsageMap::hasher().hashed_key(ns);
+    stdx::lock_guard<SimpleMutex> lk(_lock);
+
+    CollectionData& coll = _usage[hashedNs];
+    _incrementHistogram(opCtx,
+                        elapsedMicros,
+                        &coll.latencyHistogram,
+                        cmdName,
+                        isError);
+}
+
+void JoomTop::collectionDropped(const NamespaceString& nss) {
+    stdx::lock_guard<SimpleMutex> lk(_lock);
+    _usage.erase(nss.ns());
+}
+
+void JoomTop::append(BSONObjBuilder& b) {
+    stdx::lock_guard<SimpleMutex> lk(_lock);
+    _appendToUsageMap(b, _usage);
+}
+
+/* Adds statistics data to a response BSON object. Basically, it is a map of the following format:
+latencyStats: {
+    <COLLECTION_NAME> : {
+        <COMMAND_NAME> {
+            histogram: [{le: Long, count: Long}],
+            sum: Long,
+            count: Long,
+            errors: Long
+        }
+    }
+}
+*/
+void JoomTop::_appendToUsageMap(BSONObjBuilder& b, const UsageMap& map) const {
+    std::vector<std::string> names;
```

Comment:
Added preallocation, thanks. 
About maintaining of sorted vector, I am not fond of the idea, because it should be done in JoomTop::record calls and these calls are running as a part of client query execution. On other hand, it should not hurt if we sort vector on joomDiagnostics command calls, because they are relatively rare(one per 10 seconds or something like that) and they do not affect client queries because they are not part of that workload.